packages/anoncreds-rs/src/services/AnonCredsRsHolderService.ts

```diff
@@ -0,0 +1,380 @@
+        credentials.push(await credentialEntryFromAttribute(attribute))
+        credentialsProve.push({ entryIndex, isPredicate: false, referent, reveal: attribute.revealed })
+        entryIndex = entryIndex + 1
+      }
+
+      for (const referent in requestedCredentials.requestedPredicates) {
+        const predicate = requestedCredentials.requestedPredicates[referent]
+        credentials.push(await credentialEntryFromAttribute(predicate))
+        credentialsProve.push({ entryIndex, isPredicate: true, referent, reveal: true })
+        entryIndex = entryIndex + 1
+      }
+
+      // Get all requested credentials and take linkSecret. If it's not the same for every credential, throw error
+      const linkSecretsMatch = credentials.every((item) => item.linkSecretId === credentials[0].linkSecretId)
+      if (!linkSecretsMatch) {
+        throw new AnonCredsRsError('All credentials in a Proof should have been issued using the same Link Secret')
+      }
+
+      const linkSecretRecord = await agentContext.dependencyManager
+        .resolve(AnonCredsLinkSecretRepository)
+        .getByLinkSecretId(agentContext, credentials[0].linkSecretId)
+
+      if (!linkSecretRecord.value) {
+        throw new AnonCredsRsError('Link Secret value not stored')
+      }
+
+      const presentation = Presentation.create({
+        credentialDefinitions: rsCredentialDefinitions,
+        schemas: rsSchemas,
+        presentationRequest: PresentationRequest.load(JSON.stringify(proofRequest)),
+        credentials: credentials.map((entry) => entry.credentialEntry),
+        credentialsProve,
+        selfAttest: requestedCredentials.selfAttestedAttributes,
+        masterSecret: MasterSecret.load(JSON.stringify({ value: { ms: linkSecretRecord.value } })),
+      })
+
+      return JSON.parse(presentation.toJson())
+    } catch (error) {
+      agentContext.config.logger.error(`Error creating AnonCreds Proof`, {
+        error,
+        proofRequest,
+        requestedCredentials,
+      })
+      throw new AnonCredsRsError(`Error creating proof: ${error}`, { cause: error })
+    }
+  }
+
+  public async createCredentialRequest(
+    agentContext: AgentContext,
+    options: CreateCredentialRequestOptions
+  ): Promise<CreateCredentialRequestReturn> {
+    const { credentialDefinition, credentialOffer } = options
+    try {
+      const linkSecretRepository = agentContext.dependencyManager.resolve(AnonCredsLinkSecretRepository)
+
+      // If a link secret is specified, use it. Otherwise, attempt to use default link secret
+      const linkSecretRecord = options.linkSecretId
+        ? await linkSecretRepository.getByLinkSecretId(agentContext, options.linkSecretId)
+        : await linkSecretRepository.findDefault(agentContext)
+
+      if (!linkSecretRecord) {
+        // No default link secret: TODO: shall we create a new one?
+        throw new AnonCredsRsError('Link Secret not found')
```

Comment:
Yeah probably? However it can lead to race conditions if multiple request are being made at the same time 🤔